expfmt/openmetrics_create.go

```diff
@@ -0,0 +1,503 @@
+// Copyright 2014 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expfmt
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/prometheus/common/model"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// MetricFamilyToOpenMetrics converts a MetricFamily proto message into the
+// OpenMetrics text format and writes the resulting lines to 'out'. It returns
+// the number of bytes written and any error encountered. The output will have
+// the same order as the input, no further sorting is performed. Furthermore,
+// this function assumes the input is already sanitized and does not perform any
+// sanity checks. If the input contains duplicate metrics or invalid metric or
+// label names, the conversion will result in invalid text format output.
+//
+// This method fulfills the type 'prometheus.encoder'.
+//
+// Note that OpenMetrics requires a final `# EOF` line. Since this function acts
```

Comment:
I can't see where you're adding the EOF. I'd suggest adding a Close method to the Encoder interface to cover it.

Reply:
That's the point of the comment: The caller has to add the EOF. A `Close` method would be much better, but as said above, I recommend to not break users of `expfmt` right now, while we are still experimenting. When we wrap things up, we can do one bundled breaking change where we change the `Encoder` interface.

Reply:
How are you planning on doing that, considering that the caller doesn't know what has been negotiated?

Reply:
The caller does know. The flow is to first call `expfmt.Negotiate` to get the content type, and then call `expfmt.NewEncoder` with the negotiated encoding.

It's dirty, as said, but we'll fix it later:
https://github.com/prometheus/client_golang/blob/beorn7/exemplars/prometheus/promhttp/http.go#L189-L207

Reply:
Wouldn't that break anyone using this code directly, as OM will be negotiated, and then rejected by Prometheus? godoc isn't liking our org right now, so can't get an idea of impact.

I don't mind a temporary hack on our end, but let's not break people.

Reply:
Yes, that's indeed an issue. With the way the Go Modules dependency resolution works, this could even sneak in without upgrading prometheus/client_golang.

On the other hand, adding a `Close` method would not help, either. (The user that upgrades prometheus/common in the expectation of no breaking changes won't add a call of the `Close` method.) And it would break even more people (those that depend on an unchanged `Encoder` interface, for some reason).

The only way I can think of right now is to make negotiating OM an opt-in, even here on the level of prometheus/common.

Or we just keep things in a branch for now and only merge once this is finalized.

Reply:
Should we decide to leave this in a branch, I would indeed go for the `Close` method approach right now.

Reply:
> The only way I can think of right now is to make negotiating OM an opt-in, even here on the level of prometheus/common.

The plan has always been to enable negotiation by default, and ultimately make it the default if negotiation fails.

> On the other hand, adding a Close method would not help, either. 

We could change the Encode method so things would break obviously. I can't imagine there's that many direct users of this (I'd expect most would be for parsing, not generation), almost everyone should be using client_golang.

Reply:
I have tried another approach now.

I really don't want to break existing uses of this code in any way. Rather I'd keep the experiment in a branch. But I think it should work with what I have pushed now.